new/ohai_identity_attributes.md

```diff
@@ -0,0 +1,124 @@
+
+## Motivation
+
+    As a Chef user
+    I want consistent semantics for node identity attributes across platforms
+
+    As a Cookbook author
+    I want to be able to use node['fqdn'] and node['domain'] and have them DWIM
+
+    As a Cookbook author
+    I do not want to have to use attributes other than node['fqdn'] or node['domain'] to DWIM
+
+    As a Solaris user
+    I want the node['fqdn'] and node['domain'] to be computed from the short hostname and NIS domain
+
+    As a Linux user
+    I typically want to use the FQDN as my nodename and use that for derived attributes directly
+
+    As a Chef user
+    I do not want DNS failures to affect computation of my nodename/fqdn
+
+## History
+
+In Ohai 7.0.0 the semantics of the identity attributes were normalized to have the following meanings:
+
+* `hostname`:  the short hostname discovered via `hostname` (all platforms were consistent here)
+* `machinename`:  the raw output of `/bin/hostname` (this was introduced so that it was available for node
+  name resolution inside of chef-client)
+* `fqdn`:  equivalent to `/bin/hostname -f` (most platforms used this algorithm before 7.0.0, after 7.0.0 
+  it was made consistent across all platforms -- in some cases this was the return of `/bin/hostname` and
+  Solaris used hostname + NIS domain).
+* `domain`: computed from the domain suffix of the `fqdn` attribute
+
+Changing the `hostname` to be something other than the short hostname is most likely unwise since it has been
+that way consistently since it was introduced.  Because of that the `machinename` attribute was introduced to
+expose the actual return of `/bin/hostname` which is used in `node_name` guessing in the Chef::Client which
+uses the algorithm:
+
+```ruby
+name = Chef::Config[:node_name] || ohai[:fqdn] || ohai[:machinename] || ohai[:hostname]
+```
+
+It was discussed making the `fqdn` attribute to prefer the return of `/bin/hostname` when that returned a
+FQDN, but this felt to break some edge conditions in DNS related cookbooks (djbdns?) and so this was forced
+to be the fqdn which was resolved via DNS resolution across all platforms.  For platforms that do not support
+the `-f` flag to `hostname` a pure ruby version is used by ohai.
+
+The changes in Ohai 7.0.0 to the `fqdn` attribute broke Solaris people with busted DNS who expected the
+NIS domainname workaround to work.
+
+There is also a weakness in the `fqdn` attribute in that it relies on having setup forward and reverse DNS and that
+DNS is operating correctly.  A DNS failure may cause the fqdn attribute to be nil transiently.  This is likely
+a problem because Chef::Client uses it internally as first precedence for guessing the `node_name` and `fqdn` is
+used by many cookbooks.
+
+# Specification
+
+## First Proposal (Short Term Fix)
+
+The first proposal is to modify the `fqdn` attribute to be:
+
+# Equivalent to `/bin/hostname -f` if that algorithm succeeds in returning a FQDN
```

Comment:
At the moment we (I added this PR) try `/bin/hostname -f` a few times to avoid errant issues... we should keep that as part of the algorithm and then continue down the list.